components/org.wso2.carbon.identity.application.authz.opa/src/main/java/org/wso2/carbon/identity/application/authz/opa/InvokeOpaFunctionImpl.java

```diff
@@ -0,0 +1,216 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.application.authz.opa;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.conn.ConnectTimeoutException;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.wso2.carbon.identity.application.authentication.framework.AsyncProcess;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.JsGraphBuilder;
+import org.wso2.carbon.identity.application.authentication.framework.config.model.graph.js.*;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants;
+import org.wso2.carbon.identity.application.authz.opa.internal.OPAFunctionsServiceHolder;
+import org.wso2.carbon.identity.application.authz.opa.util.OPAConstants;
+import org.wso2.carbon.identity.claim.metadata.mgt.exception.ClaimMetadataException;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.LocalClaim;
+import org.wso2.carbon.identity.conditional.auth.functions.common.utils.Constants;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.*;
+
+import static org.apache.http.HttpHeaders.ACCEPT;
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
+
+/**
+ * Implementation of the InvokeOpaFunction.
+ */
+public class InvokeOpaFunctionImpl implements InvokeOpaFunction {
+
+    private static final Log LOG = LogFactory.getLog(InvokeOpaFunctionImpl.class);
+    private static final String TYPE_APPLICATION_JSON = "application/json";
+
+    private CloseableHttpClient client;
+
+    public InvokeOpaFunctionImpl() {
+
+        RequestConfig config = RequestConfig.custom()
+                .setConnectTimeout(OPAConstants.TIME)
+                .setConnectionRequestTimeout(OPAConstants.TIME)
+                .setSocketTimeout(OPAConstants.TIME)
+                .build();
+        client = HttpClientBuilder.create().setDefaultRequestConfig(config).build();
+    }
+
+    @Override
+    public void invokeOPA(String epUrl, Map<String, Object> payload, Map<String, String> options, Map<String, Object> events) {
```

Comment:
the content of this method is too long. let's break it to sub methods.

E.g. have a seperate method to sendPost, processResponse etc.